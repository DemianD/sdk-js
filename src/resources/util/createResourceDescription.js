import { HTTP_VERB_ALL_WILDCARD } from '../../constants'

export default function createResourceDescription(
  resource,
  {
    includes = [],
    filters = [],
    headers = {},
    allowed_methods = [HTTP_VERB_ALL_WILDCARD],
    deprecated = false,
  } = {},
  sub_resources = {}
) {
  return {
    resource,
    allowed_includes: includes,
    filters,
    sub_resources,
    headers,
    allowed_methods,
    deprecated,
  }
}
