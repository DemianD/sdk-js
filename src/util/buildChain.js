export default function buildChain(
  section,
  character = '.',
  chain = [section]
) {
  if (!section.includes(character)) {
    return chain
  }

  const fallback = section.substr(0, section.lastIndexOf(character))
  return buildChain(fallback, character, [...chain, fallback])
}
