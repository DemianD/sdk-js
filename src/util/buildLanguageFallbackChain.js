function buildChain(language, chain = [language]) {
  if (language.indexOf('-') === -1) {
    return chain
  }

  const fallback = language.substr(0, language.lastIndexOf('-'))
  return buildChain(fallback, [...chain, fallback])
}

export default function buildLanguageFallbackChain(language) {
  return buildChain(language)
}
