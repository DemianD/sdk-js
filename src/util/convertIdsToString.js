export default function convertIdsToString(json) {
  if (json === null) {
    return json
  }

  if (Array.isArray(json)) {
    return json.map(convertIdsToString)
  }

  if (json instanceof Object) {
    return Object.keys(json).reduce((result, key) => {
      if (
        key === 'id' ||
        (key.length > 3 && key.lastIndexOf('_id') === key.length - 3)
      ) {
        result[key] =
          json[key] !== undefined && json[key] !== null
            ? json[key].toString()
            : json[key]
      } else {
        result[key] = convertIdsToString(json[key])
      }

      return result
    }, {})
  }

  return json
}
