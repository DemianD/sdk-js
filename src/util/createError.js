export default function createError(message, type, subtype) {
  const error = new Error(message)

  // If we want we can do something with the errors in here?
  // error.stack is available

  return Object.assign(error, {
    type,
    subtype,
    internal_stack: error.stack,

    // Rewrite the stack error so that internals aren't exposed
    stack: `Error: ${message}\n\tat <SkedifyInternals>`,
  })
}
