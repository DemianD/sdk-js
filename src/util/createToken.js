import uuid from './external/uuid'

export default function createToken() {
  return uuid().replace(/-/g, '')
}
