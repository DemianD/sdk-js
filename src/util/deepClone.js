export default function deepClone(object) {
  if (!object || !Object.prototype.isPrototypeOf(object)) {
    return object
  }
  if (Array.isArray(object)) {
    return object.map(deepClone)
  }
  return Object.keys(object).reduce(
    (clone, key) => Object.assign(clone, { [key]: deepClone(object[key]) }),
    {}
  )
}
