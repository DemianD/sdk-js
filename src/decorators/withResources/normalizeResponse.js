import convertIdsToString from '../../util/convertIdsToString'

export default function normalizeResponse(response) {
  const actualResponse =
    response instanceof Error // When the response is an error
      ? response.error instanceof Error // When the response is wrapped in a retry error
        ? response.error.response
        : response.response
      : response // When the response is normal

  // Throw actual network errors
  if (!actualResponse) {
    throw new Error(response.error ? response.error.message : response.message)
  }

  const { data, warnings, errors, meta } = convertIdsToString(
    actualResponse.data
  )

  return {
    status: actualResponse.status,
    headers: actualResponse.headers,
    data,
    warnings,
    errors,
    meta,
  }
}
