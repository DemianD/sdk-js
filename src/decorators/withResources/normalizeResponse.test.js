import normalizeResponse from './normalizeResponse'
import { createRetryError } from '../../util/createError'

function createResponse(status = 200) {
  return {
    data: {
      data: 'data',
      warnings: 'warnings',
      errors: 'errors',
      meta: 'meta',
    },
    status,
    headers: [],
  }
}

describe('normalizeResponse', () => {
  it('should normalize a successful response', () => {
    expect(normalizeResponse(createResponse())).toMatchSnapshot()
  })

  it('should normalize a failed response', () => {
    expect(
      normalizeResponse(
        Object.assign(new Error('something failed'), {
          response: createResponse(),
        })
      )
    ).toMatchSnapshot()
  })

  it('should rethrow an actual a failed network request', () => {
    expect(() =>
      normalizeResponse(new Error('something failed'))
    ).toThrowErrorMatchingSnapshot()
  })

  it('should normalize a failed response after a retry', () => {
    expect(
      normalizeResponse(
        Object.assign(createRetryError(`Max retry attempts reached.`), {
          error: Object.assign(new Error('something failed'), {
            response: createResponse(),
          }),
        })
      )
    ).toMatchSnapshot()
  })

  it('should rethrow an actual a failed network request after a retry', () => {
    expect(() =>
      normalizeResponse(
        Object.assign(createRetryError(`Max retry attempts reached.`), {
          error: new Error('something failed'),
        })
      )
    ).toThrowErrorMatchingSnapshot()
  })
})
