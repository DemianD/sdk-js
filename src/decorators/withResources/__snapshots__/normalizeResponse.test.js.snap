// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeResponse should normalize a failed response 1`] = `
Object {
  "data": "data",
  "errors": "errors",
  "headers": Array [],
  "meta": "meta",
  "status": 200,
  "warnings": "warnings",
}
`;

exports[`normalizeResponse should normalize a failed response after a retry 1`] = `
Object {
  "data": "data",
  "errors": "errors",
  "headers": Array [],
  "meta": "meta",
  "status": 200,
  "warnings": "warnings",
}
`;

exports[`normalizeResponse should normalize a successful response 1`] = `
Object {
  "data": "data",
  "errors": "errors",
  "headers": Array [],
  "meta": "meta",
  "status": 200,
  "warnings": "warnings",
}
`;

exports[`normalizeResponse should rethrow an actual a failed network request 1`] = `"something failed"`;

exports[`normalizeResponse should rethrow an actual a failed network request after a retry 1`] = `"something failed"`;
